src/librustc_data_structures/obligation_forest/mod.rs

```diff
@@ -395,7 +395,16 @@ impl<O: ForestObligation> ObligationForest<O> {
         let mut errors = vec![];
         let mut stalled = true;
 
-        for index in 0..self.nodes.len() {
+        // Note that the loop body can append new nodes, and those new nodes
+        // will then be processed by subsequent iterations of the loop.
+        //
+        // We can't use an iterator for the loop because `self.nodes` is
+        // appended to and the borrow checker would complain. We also can't use
+        // `for index in 0..self.nodes.len() { ... }` because the range would
+        // be computed with the initial length, and we would miss the appended
+        // nodes. Therefore we use a `while` loop.
+        let mut index = 0;
+        while index < self.nodes.len() {
```

Comment:
Nit: What about doing this, which seems a bit less error-prone

```rust
for index in 0.. {
    // Note that the length of `self.nodes.len()` is changing, so we can't hard-code
    // a "max index" in the iterator above.
    if index >= self.nodes.len() { break; }
    ...
}
```

This avoids the need for `index += 1` in two different spots below.

Reply:
The duplicated `index += 1` bothered me, so I will do this. Thanks for the suggestion.

Reply:
I benchmarked the `for` loop and it was a 4% slowdown on `inflate-check-clean` and a 3% slowdown on `keccak-check-clean`, compared to the `while` loop.

I've seen this kind of thing before -- this `ObligationForest` code is so hot for those benchmarks that seemingly innocuous changes have an outsized effects on performance, maybe due to worse register allocation, or something like that. Anyway, I will land the original `while` loop.